libs/parallelism/algorithms/include/hpx/parallel/util/ranges_facilities.hpp

```diff
@@ -0,0 +1,61 @@
+//  Copyright (c) 2021 Giannis Gonidelis
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/concepts/concepts.hpp>
+
+#include <hpx/iterator_support/counting_iterator.hpp>
+#include <hpx/iterator_support/traits/is_sentinel_for.hpp>
+#include <hpx/parallel/algorithms/detail/advance_to_sentinel.hpp>
+
+#include <iterator>
+
+namespace hpx { namespace ranges {
+
+    // clang-format off
+    template <typename Iter,
+        HPX_CONCEPT_REQUIRES_(
+            hpx::traits::is_input_iterator<Iter>::value ||
+            hpx::traits::is_output_iterator<Iter>::value
+        )>
+    // clang-format on
+    constexpr inline Iter next(Iter first)
+    {
+        return first++;
+    }
+
+    // clang-format off
+    template <typename Iter, 
+        HPX_CONCEPT_REQUIRES_(
+            hpx::traits::is_input_iterator<Iter>::value ||
+            hpx::traits::is_output_iterator<Iter>::value
+        )>
+    // clang-format on
+    constexpr inline Iter next(
+        Iter first, typename std::iterator_traits<Iter>::difference_type dist)
+    {
+        return first + dist;
+    }
+
+    // clang-format off
+    template <typename Iter, typename Sent,
+        HPX_CONCEPT_REQUIRES_(
+            hpx::traits::is_sentinel_for<Sent, Iter>::value &&
+            hpx::traits::is_input_iterator<Iter>::value ||
+            hpx::traits::is_output_iterator<Iter>::value
```

Comment:
May I suggest to put parenthesis around the `||` part of the expression to make the intend clear?

Reply:
@hkaiser So do I go with `is_iterator` or do you prefer the parenthesis workaround?

Reply:
Either way is fine, I guess.